src/backend/impl/src/repositories/types/commit_sha.rs

```diff
@@ -0,0 +1,178 @@
+use std::borrow::Cow;
+
+use backend_api::ApiError;
+use candid::{
+    types::{Serializer, Type, TypeInner},
+    CandidType, Deserialize,
+};
+use hex::FromHex;
+use ic_stable_structures::{storable::Bound, Storable};
+
+const COMMIT_SHA_SIZE: usize = 20;
+
+type CommitShaInner = [u8; COMMIT_SHA_SIZE];
+
+#[derive(Debug, Clone, Copy, Default, Ord, PartialOrd, PartialEq, Eq)]
+pub struct CommitSha(CommitShaInner);
+
+impl CommitSha {
+    const MIN: CommitSha = CommitSha([0; COMMIT_SHA_SIZE]);
+    const MAX: CommitSha = CommitSha([255; COMMIT_SHA_SIZE]);
+
+    pub fn new(bytes: CommitShaInner) -> Self {
+        Self(bytes)
+    }
+
+    pub fn min() -> Self {
+        Self::MIN
+    }
+
+    pub fn max() -> Self {
+        Self::MAX
+    }
+}
+
+impl TryFrom<&str> for CommitSha {
+    type Error = ApiError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        if value.len() != COMMIT_SHA_SIZE * 2 {
+            return Err(ApiError::internal(&format!(
+                "Invalid commit sha length: {}",
+                value.len()
+            )));
+        }
+
+        let bytes = CommitShaInner::from_hex(value)
+            .map_err(|e| ApiError::internal(&format!("Failed to decode commit sha: {:?}", e)))?;
+
+        Ok(Self(bytes))
+    }
+}
+
+impl TryFrom<Vec<u8>> for CommitSha {
+    type Error = ApiError;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        if value.len() != COMMIT_SHA_SIZE {
```

Comment:
This will force reviewers to submit the long commit hashes. Normally the short commit hashes are used in the proposals so reviewers will have to go get the long commit hash, which would be an additional step that they don't have to do now.

In the longer term we can try to automate the conversion of short commit hash to long commit hash so that the long one is always submitted even if the short one is entered by the user, but we need the full context of the git repo in the browser before we can do that.

In the meantime, I think we have a few options:
- Keep it as it is, but force reviewers to enter long commit hash
- Drop the requirement for one review per commit hash and hope that reviewers don't submit duplicate commit hashes
- Drop the commit hash from the index and just do a less efficient, full table scan for duplicates 

I think I like the 3rd option because the performance should be okay for awhile and we can hopefully start on the git automation before performance becomes a problem. What do you think?

Reply:
I like the third option, but it's not easy to find duplicates if the short commit hash length can vary. We have to be careful when scanning for duplicates and have a more advanced check on the commit hashes than just the equality.

Reply:
We agreed to use the GitHub API on the UI to convert short hashes into long hashes. e.g. https://api.github.com/repos/dfinity/ic/commits/e98988c08